File: Abstract Factory Design Pattern/UI/C++/ui.cpp
#include <string>
#include <iostream>
using namespace std;

class Ibutton{
    public :
    virtual void press() = 0;
   
};
class ITextbox{
    public :
    virtual void settext() = 0;
   
};

class MacButton : public Ibutton{
    public :
    void press(){
        cout<<"Mac button pressed"<<endl;
    }
   
};

class WinButton : public Ibutton{
    public :
    void press(){
        cout<<"Win button pressed"<<endl;
    }
   
};

class MacTextBox : public ITextbox{
    public :
    void settext(){
        cout<<"Setting text in Mac Textbox"<<endl;
    }
   
};

class WinTextBox : public ITextbox{
    public :
    void settext(){
        cout<<"Setting text in Win Textbox"<<endl;
    }
   
};

class IFactory{
    public :
        virtual Ibutton* createbutton() = 0;
        virtual ITextbox* createtextbox() = 0;
};


class WinFactory : public IFactory{
    public :
        Ibutton* createbutton(){
            return new WinButton();
        }
        ITextbox* createtextbox(){
            return new WinTextBox();
        }
   
};

class MacFactory : public IFactory{
    public :
    Ibutton* createbutton(){
        return new MacButton();
    }
    ITextbox* createtextbox(){
        return new MacTextBox();
    }
   
};

class GUIAbstractfactory{
  public:
    static IFactory* createfactory(string ostype){
        if(ostype == "windows"){
            return new WinFactory();
        }
        else if(ostype == "mac"){
            return new MacFactory;
        }
        return nullptr;
    }
};


int main(){
   
    cout<<"Enter machine OS\n";
    string ostype;
    cin>>ostype;
   
    IFactory* fact = GUIAbstractfactory::createfactory(ostype);
   
    Ibutton* button = fact->createbutton();
    button->press();
   
    ITextbox* textBox = fact->createtextbox();
    textBox->settext();
   
    return 0;
}

File: Abstract Factory Design Pattern/UI/JAVA/UI.java
import java.util.Scanner;

interface IButton {
    void press();
}

interface ITextbox {
    void settext();
}

class MacButton implements IButton {
    @Override
    public void press() {
        System.out.println("Mac button pressed");
    }
}

class WinButton implements IButton {
    @Override
    public void press() {
        System.out.println("Win button pressed");
    }
}

class MacTextBox implements ITextbox {
    @Override
    public void settext() {
        System.out.println("Setting text in Mac Textbox");
    }
}

class WinTextBox implements ITextbox {
    @Override
    public void settext() {
        System.out.println("Setting text in Win Textbox");
    }
}

interface IFactory {
    IButton createButton();
    ITextbox createTextbox();
}

class WinFactory implements IFactory {
    @Override
    public IButton createButton() {
        return new WinButton();
    }

    @Override
    public ITextbox createTextbox() {
        return new WinTextBox();
    }
}

class MacFactory implements IFactory {
    @Override
    public IButton createButton() {
        return new MacButton();
    }

    @Override
    public ITextbox createTextbox() {
        return new MacTextBox();
    }
}

class GUIAbstractFactory {
    public static IFactory createFactory(String osType) {
        if (osType.equals("windows")) {
            return new WinFactory();
        } else if (osType.equals("mac")) {
            return new MacFactory();
        }
        
        return null;
    }
}

public class UI {
    public static void main(String[] args) {
        System.out.println("Enter machine OS");
        Scanner scanner = new Scanner(System.in);
        String osType = scanner.nextLine();
        scanner.close();

        IFactory factory = GUIAbstractFactory.createFactory(osType);

        if (factory != null) {
            IButton button = factory.createButton();
            button.press();

            ITextbox textBox = factory.createTextbox();
            textBox.settext();
        } else {
            System.out.println("Invalid OS type");
        }
    }
}
